clusterctl/clusterdeployer/clusterdeployer.go

```diff
@@ -3,36 +3,313 @@ package clusterdeployer
+	}
+
+	glog.V(2).Info("Pivoting Cluster API objects from eternal to internal cluster.")
+	err = pivot(externalClient, internalClient)
+	if err != nil {
+		return fmt.Errorf("unable to pivot cluster API objects: %v", err)
+	}
+
+	glog.V(2).Info("Applying Cluster API provider components to internal cluster.")
+	err = d.ApplyClusterAPIControllers(internalClient)
+	if err != nil {
+		return fmt.Errorf("unable to apply cluster api provider components to internal cluster:%v", err)
+	}
+
+	// Update cluster endpoint as the status does not always move.
+	// Needed till this logic moves into cluster controller.
+	// TODO: https://github.com/kubernetes-sigs/cluster-api/issues/158
+	err = internalClient.UpdateClusterObjectEndpoint(masterIP)
+	if err != nil {
+		return fmt.Errorf("unable to update cluster endpoint: %v", err)
+	}
+
+	glog.V(2).Info("Creating node machines in internal cluster.")
+	err = internalClient.CreateMachineObjects(nodes)
+	if err != nil {
+		return fmt.Errorf("unable to create node machines: %v", err)
+	}
+
+	return nil
+}
+
+func (d *ClusterDeployer) ApplyClusterAPIApiserver(client ClusterClient) error {
+	yaml, err := GetApiServerYaml()
+	if err != nil {
+		return fmt.Errorf("unable to generate apiserver yaml:%v", err)
+	}
+
+	err = client.Apply(yaml)
+	if err != nil {
+		return fmt.Errorf("unable to apply apiserver yaml:%v", err)
+	}
+	return client.WaitForClusterV1alpha1Ready()
+}
+
+func (d *ClusterDeployer) ApplyClusterAPIControllers(client ClusterClient) error {
+	return client.Apply(d.providerComponents)
+}
+
+func (d *ClusterDeployer) writeKubeconfig(kubeconfig string) error {
+	os.Remove(d.kubeconfigOutput)
+	f, err := os.Create(d.kubeconfigOutput)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	f.WriteString(kubeconfig)
+	return nil
+}
+
+func waitForKubeconfigReady(provider ProviderDeployer, machine *clusterv1.Machine) (string, error) {
+	kubeconfig := ""
+	err := util.Poll(500*time.Millisecond, 120*time.Second, func() (bool, error) {
+		glog.V(2).Infof("Waiting for kubeconfig on %v to become ready...", machine.Name)
+		k, err := provider.GetKubeConfig(machine)
+		if err != nil {
+			glog.V(4).Infof("error getting kubeconfig: %v", err)
+			return false, nil
+		}
+		if k == "" {
+			return false, nil
+		}
+		kubeconfig = k
+		return true, nil
+	})
+
+	return kubeconfig, err
+}
+
+func pivot(from, to ClusterClient) error {
+	if err := from.WaitForClusterV1alpha1Ready(); err != nil {
+		return fmt.Errorf("Cluster v1aplpha1 resource not ready on source cluster.")
+	}
+
+	if err := to.WaitForClusterV1alpha1Ready(); err != nil {
+		return fmt.Errorf("Cluster v1aplpha1 resource not ready on target cluster.")
+	}
+
+	// TODO: As we move objects, update any references (eg. owner ref) in following object to new UID
+	clusters, err := from.GetClusterObjects()
+	if err != nil {
+		return err
+	}
+
+	for _, cluster := range clusters {
+		cluster.SetResourceVersion("")
```

Comment:
can you comment on why it is necessary to set the resource version to an empty string

Reply:
Fixed.